src/main/java/com/ditrit/letomodelizerapi/service/LibraryServiceImpl.java

```diff
@@ -0,0 +1,359 @@
+        if (libraryWhitelistHosts.stream().noneMatch(host -> libraryRecord.url().startsWith(host))) {
+            throw new ApiException(ErrorType.UNAUTHORIZED_LIBRARY_URL, "url", libraryRecord.url());
+        }
+
+        AccessControl role = null;
+        if (StringUtils.isNotBlank(libraryRecord.role())) {
+            role = accessControlService.create(AccessControlType.ROLE, new AccessControlRecord(libraryRecord.role()));
+            accessControlService.associateUser(AccessControlType.ROLE, role.getId(), login);
+        }
+
+        String libraryUrl = libraryRecord.url().replace("index.json", "");
+
+        if (libraryRepository.existsByUrl(libraryUrl)) {
+            throw new ApiException(ErrorType.ENTITY_ALREADY_EXISTS, "url", libraryUrl);
+        }
+
+        Library library = save(libraryUrl, null);
+
+        permissionService.createLibraryPermissions(library, role);
+
+        return library;
+    }
+
+    @Override
+    public void update(final Long id, final String url) throws JsonProcessingException {
+        findById(id);
+        if (libraryWhitelistHosts.stream().noneMatch(url::startsWith)) {
+            throw new ApiException(ErrorType.UNAUTHORIZED_LIBRARY_URL, "url", url);
+        }
+
+        String libraryUrl = url.replace("index.json", "");
+
+        if (libraryRepository.existsByUrlAndIdIsNot(libraryUrl, id)) {
+            throw new ApiException(ErrorType.ENTITY_ALREADY_EXISTS, "url", libraryUrl);
+        }
+
+        libraryTemplateRepository.deleteByLibraryId(id);
+        save(libraryUrl, id);
+    }
+
+    /**
+     * Saves a library entity with the provided URL, potentially updating an existing library if an ID is provided.
+     * This method downloads the library data from the specified URL, validates the JSON structure, and then
+     * creates or updates a library entity based on the JSON data. It also processes any associated library templates
+     * included in the library's JSON data, saving them to the repository.
+     *
+     * @param url the URL from which to download the library data, appended with "index.json" to target the specific
+     *            JSON file
+     * @param id the optional ID of the library to update; if null, a new library entity is created
+     * @return the saved or updated Library entity
+     * @throws JsonProcessingException if there is an error reading the JSON data into a JsonNode
+     * Note: JsonProcessingException is mentioned in the method signature for clarity, but as noted, it cannot actually
+     * be thrown due to prior validation of the JSON data.
+     */
+    Library save(final String url, final Long id) throws JsonProcessingException {
+        String libraryValue = downloadLibrary(String.format("%sindex.json", url));
+        validate(libraryValue);
+
+        // JsonProcessingException can't be thrown because, we already validate json before.
```

Comment:
```suggestion
        // JsonProcessingException can't be thrown because we already validate json before.
```

Or if you really want to add an `,`, put it here `thrown, because...`

Reply:
done